tests/src/test/java/org/eclipse/hono/tests/client/ClientTestBase.java

```diff
@@ -230,7 +193,29 @@ private void disconnect(final TestContext ctx) {
      * @param messageConsumer The handler to invoke for every message received.
      * @return A future succeeding with the created consumer.
      */
-    abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
+    protected abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
+
+    /**
+     * Prepare the downstream consumer client for connection to the messaging network.
+     * 
+     * @return The client for connecting to the messaging network.
+     */
+    protected abstract HonoClient prepareDownstreamClient();
+
+    /**
+     * Prepares the AMQP 1.0 client for connecting to one of Hono's services.
+     * 
+     * @param ctx The Vert.x test context.
+     * 
+     * @return The client for connecting to a Hono service.
+     */
+    protected abstract HonoClient prepareHonoClient(TestContext ctx);
+
+    protected abstract Future<Void> registerDevice();
```

Comment:
I would rather create and register random tenants and devices as part of the individual tests (e.g. as done in the HTTP and MQTT adapter ITs).

Reply:
ok

Reply:
@sophokles73 am having an issue with this proposition i.e creating a random tenant for the hono messaging tests. during the test, clients connect to the messaging service using the user `hono-client`, however, this user is only authorize to send events/telemetry for the default tenant.

this is different though for the protocol adapters -> they are authorize to publish telemetry/events for any tenant (so generating random tenantIds works in this case).

Can i change the permissions on record for the `hono-client` user so that I can create a sender for a tenant? WDYT?

Reply:
or maybe generate random tenants only for the amqp adapter and use the default tenant for the hono-messaging tests?

Reply:
On a second thought, we could also use one of the protocol adapter users for the messaging tests, couldn't we?

Reply:
yes sure. i will try that in the next commit